mantidproject/mantid
# Enable other focussing modes for tt_mode=custom in PEARL powder diff scripts

**Description of work.**

A new tt_mode "custom" was added to the PEARL powder diffraction scripts in https://github.com/mantidproject/mantid/pull/31841. This only supported the focusing mode "Mods" due to the fact any grouping file could be used and it wasn't certain there would be the 14 standard groups

This change checks if there are 14 groups and if there are the other focusing modes are available. Also if the code defaults to focusing mode "mods" it writes a warning saying this is what has happened

I also added in a system test for tt_mode=custom. Since this test suite was starting to take a long time I've speeded it up by removing the history from the V spline input file (PEARL00098472_splined.nxs) and also a couple of the ref files. The time to load the history meant they were taking 40s+ to load each time.

**To test:**

Rerun the script from https://github.com/mantidproject/mantid/pull/31841. This uses a grouping file with a single group in it and has focus_mode="all" set in one of the yaml files. Note that it now produces a warning says there weren't 14 groups in the supplied grouping file so it is defaulting to focus mode "Mods". The output workspace group has a single workspace in it.

Change the `custom_grouping_file` parameter in the calls to create_vanadium and focus to "pearl_group_12_1_TT70.cal" and observe that there is no warning now (because this does has 14 groups). The output workspace group contains a workspace for mods1-9 and 5 more workspaces for mods 10-14 which is the standard output for focus_mode=all.

Fixes #32956.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
